File: packages/javascript/src/resources/knowledge.ts
/**
 * Knowledge resource — knowledge bases, documents, search, library, attachments.
 */

import { del, get, getList, post, put, request, uploadFile } from "../http.js";
import type {
  KnowledgeBase,
  KnowledgeBaseCreate,
  KnowledgeBaseUpdate,
  KnowledgeDocument,
  ResolvedConfig,
  SearchResponse,
  SearchResult,
  WorkforceKnowledgeBaseLink,
} from "../types.js";

export class KnowledgeResource {
  private readonly config: ResolvedConfig;

  constructor(config: ResolvedConfig) {
    this.config = config;
  }

  // ── KB CRUD ───────────────────────────────────────────

  /** List every knowledge base in the caller's library. */
  listAll(): Promise<KnowledgeBase[]> {
    return getList<KnowledgeBase>(this.config, "/knowledge-bases");
  }

  /** List featured / templated knowledge bases. */
  listFeatured(): Promise<KnowledgeBase[]> {
    return getList<KnowledgeBase>(this.config, "/knowledge-bases/featured");
  }

  /**
   * List knowledge bases attached to a workforce.
   *
   * Returns the inner {@link KnowledgeBase} from each attachment record.
   */
  async list(workforceId: string): Promise<KnowledgeBase[]> {
    type AttachItem = { knowledge_base?: KnowledgeBase };
    const payload = await request<AttachItem[] | null>(
      this.config,
      "GET",
      `/workforces/${workforceId}/knowledge-bases`,
    );
    if (!Array.isArray(payload)) {
      return [];
    }
    return payload
      .filter((item): item is { knowledge_base: KnowledgeBase } =>
        Boolean(item && item.knowledge_base),
      )
      .map((item) => item.knowledge_base);
  }

  /** Get a single knowledge base by ID. */
  get(knowledgeBaseId: string): Promise<KnowledgeBase> {
    return get<KnowledgeBase>(
      this.config,
      `/knowledge-bases/${knowledgeBaseId}`,
    );
  }

  /** Create a knowledge base (library or workforce-scoped). */
  create(data: KnowledgeBaseCreate): Promise<KnowledgeBase> {
    return post<KnowledgeBase>(this.config, "/knowledge-bases", data);
  }

  /** Update a knowledge base (server uses PUT). */
  update(
    knowledgeBaseId: string,
    data: KnowledgeBaseUpdate,
  ): Promise<KnowledgeBase> {
    return put<KnowledgeBase>(
      this.config,
      `/knowledge-bases/${knowledgeBaseId}`,
      data,
    );
  }

  /** Delete a knowledge base. */
  delete(knowledgeBaseId: string): Promise<void> {
    return del<void>(this.config, `/knowledge-bases/${knowledgeBaseId}`);
  }

  // ── Workforce attachment ──────────────────────────────

  /** Attach an existing KB to a workforce. */
  attachToWorkforce(
    workforceId: string,
    knowledgeBaseId: string,
  ): Promise<WorkforceKnowledgeBaseLink> {
    return post<WorkforceKnowledgeBaseLink>(
      this.config,
      `/workforces/${workforceId}/knowledge-bases`,
      { knowledge_base_id: knowledgeBaseId },
    );
  }

  /** Detach a KB from a workforce. */
  detachFromWorkforce(
    workforceId: string,
    knowledgeBaseId: string,
  ): Promise<void> {
    return del<void>(
      this.config,
      `/workforces/${workforceId}/knowledge-bases/${knowledgeBaseId}`,
    );
  }

  // ── Documents ─────────────────────────────────────────

  /**
   * Upload a document to a knowledge base.
   *
   * In browsers, pass a `File`. In Node.js 18+, use `File`/`Blob` from
   * `node:buffer`.
   */
  uploadDocument(
    knowledgeBaseId: string,
    file: Blob | File,
    fileName?: string,
  ): Promise<KnowledgeDocument> {
    return uploadFile<KnowledgeDocument>(
      this.config,
      `/knowledge-bases/${knowledgeBaseId}/documents`,
      "file",
      file,
      fileName,
    );
  }

  /** Download the raw bytes for a document. */
  async downloadDocument(
    knowledgeBaseId: string,
    documentId: string,
  ): Promise<ArrayBuffer> {
    const url = new URL(
      `/api/v1/knowledge-bases/${knowledgeBaseId}/documents/${documentId}/download`,
      this.config.baseUrl,
    ).toString();
    const headers: Record<string, string> = {};
    if (this.config.apiKey) {
      headers["X-API-Key"] = this.config.apiKey;
    }
    const response = await fetch(url, {
      method: "GET",
      headers,
      signal: AbortSignal.timeout(this.config.timeout),
    });
    if (!response.ok) {
      throw new Error(
        `Document download failed: ${response.status} ${response.statusText}`,
      );
    }
    return response.arrayBuffer();
  }

  /** Delete a document from a knowledge base. */
  deleteDocument(knowledgeBaseId: string, documentId: string): Promise<void> {
    return del<void>(
      this.config,
      `/knowledge-bases/${knowledgeBaseId}/documents/${documentId}`,
    );
  }

  // ── Search ────────────────────────────────────────────

  /**
   * Run a semantic / keyword search against a KB.
   *
   * Returns the result list directly. Use {@link searchResponse} for the
   * full envelope (including `query` and `total`).
   */
  async search(
    knowledgeBaseId: string,
    query: string,
    options: { topK?: number; filters?: Record<string, unknown> } = {},
  ): Promise<SearchResult[]> {
    const envelope = await this.searchResponse(knowledgeBaseId, query, options);
    return envelope.results;
  }

  /** Run a search and return the full {@link SearchResponse} envelope. */
  searchResponse(
    knowledgeBaseId: string,
    query: string,
    options: { topK?: number; filters?: Record<string, unknown> } = {},
  ): Promise<SearchResponse> {
    const body: Record<string, unknown> = {
      query,
      top_k: options.topK ?? 5,
    };
    if (options.filters !== undefined) {
      body["filters"] = options.filters;
    }
    return post<SearchResponse>(
      this.config,
      `/knowledge-bases/${knowledgeBaseId}/search`,
      body,
    );
  }
}

File: packages/python/navaia_forge/resources/knowledge.py
"""Knowledge base resource (KBs, documents, search, library)."""

from __future__ import annotations

from typing import Any

from ..types import (
    KnowledgeBase,
    KnowledgeSourceType,
    SearchResponse,
    SearchResult,
)
from ._base import ResourceBase, parse_list, parse_model


class KnowledgeResource(ResourceBase):
    """Operations for knowledge bases and search."""

    # ── KB CRUD ───────────────────────────────────────────────

    def list_all(self) -> list[KnowledgeBase]:
        """List every knowledge base in the caller's library."""
        return parse_list(KnowledgeBase, self._http.get_list("/knowledge-bases"))

    def list(self, workforce_id: str) -> list[KnowledgeBase]:
        """List the knowledge bases belonging to a workforce.

        The backend scopes KBs by ``workforce_id`` on the collection endpoint
        (``GET /knowledge-bases?workforce_id=...``); the param is required.
        """
        return parse_list(
            KnowledgeBase,
            self._http.get_list(
                "/knowledge-bases", params={"workforce_id": workforce_id}
            ),
        )

    def get(self, knowledge_base_id: str) -> KnowledgeBase:
        """Fetch a knowledge base by id."""
        return parse_model(
            KnowledgeBase,
            self._http.get(f"/knowledge-bases/{knowledge_base_id}"),
        )

    def create(
        self,
        name: str,
        *,
        description: str = "",
        workforce_id: str | None = None,
        source_type: KnowledgeSourceType = "upload",
        retrieval_mode: str = "semantic",
        config_json: dict[str, Any] | None = None,
    ) -> KnowledgeBase:
        """Create a knowledge base (library or workforce-scoped)."""
        body: dict[str, Any] = {
            "name": name,
            "description": description,
            "retrieval_mode": retrieval_mode,
            "source_type": source_type,
        }
        if workforce_id is not None:
            body["workforce_id"] = workforce_id
        if config_json is not None:
            body["config_json"] = config_json
        return parse_model(KnowledgeBase, self._http.post("/knowledge-bases", body))

    def update(self, knowledge_base_id: str, **fields: Any) -> KnowledgeBase:
        """Update a knowledge base (server uses PATCH)."""
        return parse_model(
            KnowledgeBase,
            self._http.patch(f"/knowledge-bases/{knowledge_base_id}", fields),
        )

    def delete(self, knowledge_base_id: str) -> None:
        """Delete a knowledge base."""
        self._http.delete(f"/knowledge-bases/{knowledge_base_id}")

    # ── Search ────────────────────────────────────────────────

    def search(
        self,
        knowledge_base_id: str,
        query: str,
        *,
        top_k: int = 5,
        filters: dict[str, Any] | None = None,
    ) -> list[SearchResult]:
        """Run a semantic / keyword search against a KB.

        Returns the result list directly (use :meth:`search_response` for
        the full envelope including ``query`` and ``total``).
        """
        return self.search_response(
            knowledge_base_id, query, top_k=top_k, filters=filters
        ).results

    def search_response(
        self,
        knowledge_base_id: str,
        query: str,
        *,
        top_k: int = 5,
        filters: dict[str, Any] | None = None,
    ) -> SearchResponse:
        """Run a search and return the full :class:`SearchResponse` envelope."""
        body: dict[str, Any] = {"query": query, "top_k": top_k}
        if filters is not None:
            body["filters"] = filters
        return parse_model(
            SearchResponse,
            self._http.post(f"/knowledge-bases/{knowledge_base_id}/search", body),
        )
